File: TOP10majorSorting/countsort.py
def count_sort(nums):
    minval=min(nums)
    maxval=max(nums)
    rangeval=maxval-minval
    count =[0]*(rangeval+1)
    for num in nums:
        count[num-minval]+=1
    sortans=[]
    for id in range(rangeval+1):
        val=id+minval
        sortans.extend([val]*count[id])
    return sortans 
    
nums=[8,-4,2,-1,0]
ans= count_sort(nums.copy())
print(ans)

File: TOP10majorSorting/countsort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
void countsort(vector<int> &nums)
{
    int minval = *min_element(nums.begin(), nums.end());
    int maxval = *max_element(nums.begin(), nums.end());
    int valrange = maxval - minval;
    vector<int> count(valrange + 1, 0);
    for (int num : nums)
    {
        count[num - minval] += 1;
    }
    int id = 0;
    for (int i = 0; i < count.size(); i++)
    {
        int val = i + minval;
        while (count[i] > 0)
        {
            nums[id++] = val;
            count[i]--;
        }
    }
}
int main(void)
{
    vector<int> nums = {8, -4, 2, -1, 0};
    countsort(nums);
    for (int num : nums)
    {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}